### ISSUE TITLE
Leverage NativeTypedef rather than adding AlsoUsableFor?

### ISSUE THREAD

**Issue by marler8997** [microsoft/win32metadata#398]
It seems to me that the relationship between a type and its `AlsoUsableFor` type is the same relationship between a type and its `NativeTypedef` target.  If this is true, then we could leverage the existing `NativeTypedef` mechanism to express this relationship.  So instead of this:

```C#
[AlsoUsableFor("HeapHandle")]
[NativeTypedef]
public struct ProcessHeapHandle
{
	public IntPtr Value;
}
```

The metadata could do this:
```C#
[NativeTypedef]
public struct ProcessHeapHandle
{
	public HeapHandle Value;
}
```

**Comment by AArnott** [comment]
That may be true, assuming there is never a need for one struct to be usable for multiple other structs.

But since we already have a solution for this, why do you feel it's worth re-solving it another way? Projections have likely already been built around the way the metadata has addressed this requirement so we'd need a compelling benefit to go to the work to solving this problem over again.

**Comment by kennykerr** [comment]
This is also not compatible with C++ or Rust where the outer struct would destroy the nested struct, leading to incorrect behavior.

**Comment by marler8997** [comment]
> But since we already have a solution for this, why do you feel it's worth re-solving it another way?

Just seems like having 1 way to do things is better than 2.

> This is also not compatible with C++ or Rust where the outer struct would destroy the nested struct, leading to incorrect behavior.

Could you elaborate?

**Comment by kennykerr** [comment]
Well if the nested type has a destructor, like `HeapHandle`, then the containing type would also be given a destructor according to the rules of those languages. But `ProcessHeapHandle` is not meant to be destroyed - it is only meant to be usable in places where `HeapHandle` is used.

**Comment by marler8997** [comment]
@kennykerr I'm not sure I understand, `NativeTypedef` currently has no target types with destructors correct?  They are currently all native types like `IntPtr` and `Int32` and `const void*` etc.  So how would you have functionality that causes destructors to be inherited if no such use case exists?  Do you have a use case where this happens?

Could you please reopen this issue until discussion has finished?  I'm not sure the issue has been resolved yet.  If you'll allow me, I'll close it when it seems like its' been resolved.

**Comment by kennykerr** [comment]
Sure, have a look at the metadata for `Windows.Win32.SystemServices.HeapHandle`. It includes the `RAIIFree` attribute, which implies a destructor will be generated (for applicable languages).

**Comment by marler8997** [comment]
I got an idea to try to understand what C++/Rust might be doing that would help me understand.  Based on another comment of yours, I learned that those projections translate `NativeTypedef` into a struct wrapper.  I'm not familair with Rust but in C++ I believe this translation would automatically cause the field's destructor to be included in the new wrapper type's default constructor. So based on your latest comment, it sounds like you're NativeTypedef wrappers use the default constructor and thus, would inherit the destructor.  So then the only part I don't understand which I would be quite interested to know is what do those projections translate `AlsoUsableFor` into?
